packages/engine/lib/error/src/future.rs

```diff
@@ -18,213 +18,124 @@ use pin_project::pin_project;
-    message: Option<M>,
-}
-
-impl<Fut, M> Future for FutureWithMessage<Fut, M>
-where
-    Fut: Future,
-    Fut::Output: ResultExt,
-    M: fmt::Display + fmt::Debug + Send + Sync + 'static,
-{
-    type Output = Fut::Output;
-
-    #[track_caller]
-    fn poll(self: Pin<&mut Self>, cx: &mut TaskContext) -> Poll<Self::Output> {
-        let projection = self.project();
-        let inner = projection.inner;
-        let message = projection.message;
-
-        // Can't use `map` as `#[track_caller]` is unstable on closures
-        match inner.poll(cx) {
-            Poll::Ready(value) => Poll::Ready(value.attach_message({
-                message
-                    .take()
-                    .expect("Cannot poll context after it resolves")
-            })),
-            Poll::Pending => Poll::Pending,
+macro_rules! implement_future_adaptor {
+    ($future:ident, $method:ident, $bound:ident $(+ $bounds:ident)* $(+ $lifetime:lifetime)*, $output:ty) => {
+        #[doc = concat!("Adaptor returned by [`FutureExt::", stringify!( $method ), "`].")]
+        #[pin_project]
+        pub struct $future<Fut, T> {
+            #[pin]
+            future: Fut,
+            inner: Option<T>,
         }
-    }
-}
-
-/// Adaptor returned by [`FutureExt::attach_message_lazy`].
-#[pin_project]
-pub struct FutureWithLazyMessage<Fut, F> {
-    #[pin]
-    inner: Fut,
-    op: Option<F>,
-}
-
-impl<Fut, F, M> Future for FutureWithLazyMessage<Fut, F>
-where
-    Fut: Future,
-    Fut::Output: ResultExt,
-    F: FnOnce() -> M,
-    M: fmt::Display + fmt::Debug + Send + Sync + 'static,
-{
-    type Output = Fut::Output;
 
-    #[track_caller]
-    fn poll(self: Pin<&mut Self>, cx: &mut TaskContext) -> Poll<Self::Output> {
-        let projection = self.project();
-        let inner = projection.inner;
-        let op = projection.op;
-
-        // Can't use `map` as `#[track_caller]` is unstable on closures
-        match inner.poll(cx) {
-            Poll::Ready(value) => Poll::Ready(
-                value
-                    .attach_message_lazy(op.take().expect("Cannot poll context after it resolves")),
-            ),
-            Poll::Pending => Poll::Pending,
+        impl<Fut, T> Future for $future<Fut, T>
+        where
+            Fut: Future,
+            Fut::Output: ResultExt,
+            T: $bound $(+ $bounds)* $(+ $lifetime)*
+        {
+            type Output = $output;
+
+            #[track_caller]
+            fn poll(self: Pin<&mut Self>, cx: &mut TaskContext) -> Poll<Self::Output> {
+                let projection = self.project();
+                let future = projection.future;
+                let inner = projection.inner;
+
+                // Can't use `map` as `#[track_caller]` is unstable on closures
+                match future.poll(cx) {
+                    Poll::Ready(value) => {
+                        Poll::Ready(value.$method({
+                            inner.take().expect("Cannot poll context after it resolves")
```

Comment:
Are these panic messages correct now? (appreciate they were the same before but perhaps incorrect then as well)

Reply:
Yes, it's not allowed to poll a Future after its polling has returned `Poll::Ready`, thus the panic message is correct

Reply:
I mean the fact that it says `context`

Reply:
Yes, the context is related to `TaskContext` which is imported as `use task::Context as TaskContext`. It's a little bit unfortunate, that we use the same term, but in this case, we need to stick with `context`.

Reply:
Makes sense